### PR TITLE
#3847 Store Page Editor validation results in Redux instead of Formik

### LINKED ISSUE #3847: Page editor validation improvements
1. Do not run pipeline validation (Trace errors, renderer validation, etc) on every change of a block.
2. Persist validation results between nodes and elements switches.

### PR DESCRIPTION
**Pull request by BALEHOK**
## What does this PR do?

- Closes #3847 
- Stores the extension error state in Redux
- Moves template validation to the field level
- Works for block-level trace errors and form-level Formik validation errors

## Reviewer Tips

- Main focus is the approach itself: store errors in Redux on edit, apply stored errors to the Formik context when a node is selected

## Demo

The error "expected absolute URL for request without service" in the video is generated from the traces, applied to the block, and stored in Redux. The text template validation is triggered by the field, set in the Formik context, and propagated to the Redux state; when the extension is selected next time, the error state is loaded from Redux and sent to Formik.

https://www.loom.com/share/61f5d2a5718244dfa72f126e92f19f73

## Checklist

- [ ] Add tests
- [x] Designate a primary reviewer @BLoe

### CODE REVIEW

File: src/pageEditor/hooks/useDebouncedEffect.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { useEffect, useState } from "react";
+import { useDebounce } from "use-debounce";
+import { isEqual } from "lodash";
+
+const useDebouncedEffect = (
```
Comment (BALEHOK) [inline src/pageEditor/hooks/useDebouncedEffect.ts:22]: This is the `Effect` component converted to hook.

File: src/pageEditor/hooks/usePipelineErrors.ts [hunk 2]
```diff
@@ -61,6 +69,42 @@ function usePipelineErrors() {
     validate: validatePipelineBlocks,
   });
 
+  useEffect(() => {
```
Comment (BALEHOK) [inline src/pageEditor/hooks/usePipelineErrors.ts:72]: This replaces `applyTraceBlockError`; it operated on the block (node) level, sets the error directly to Redux state, and doesn't touch Formik, decoupling Trace errors and Formik (`validateRenderers` and `trace input error` will need to be refactored in a similar fashion).

File: src/pageEditor/slices/editorSlice.ts [hunk 3]
```diff
@@ -653,10 +658,8 @@ export const editorSlice = createSlice({
     },
     removeNode(state, action: PayloadAction<UUID>) {
       const nodeIdToRemove = action.payload;
-      const element = state.elements.find(
-        (x) => x.uuid === state.activeElementId
-      );
-      const elementUiState = state.elementUIStates[state.activeElementId];
+      const element = selectActiveElement({ editor: state });
+      const elementUiState = selectActiveElementUIState({ editor: state });
```
Comment (BALEHOK) [inline src/pageEditor/slices/editorSlice.ts:662]: Re-using selectors in the reducer

File: src/pageEditor/slices/editorSlice.ts [hunk 4]
```diff
@@ -678,6 +681,27 @@ export const editorSlice = createSlice({
       // This change should re-initialize the Page Editor Formik form
       state.selectionSeq++;
     },
+    setError(
+      state,
+      action: PayloadAction<{
+        nodeId?: UUID;
+        nodeError?: string;
+        fieldErrors?: FormikErrorTree;
```
Comment (BALEHOK) [inline src/pageEditor/slices/editorSlice.ts:689]: Form error state, synced with Formik on the node level (not for the whole extension)

File: src/pageEditor/slices/editorSlice.ts [hunk 5]
```diff
@@ -678,6 +681,27 @@ export const editorSlice = createSlice({
...
+        action.payload.nodeId ?? selectActiveNodeId({ editor: state });
+      const elementUiState = selectActiveElementUIState({ editor: state });
+      const elementErrors = elementUiState.errorMap;
+      if (typeof nodeError !== "undefined") {
+        set(elementErrors, [nodeId, "message"], nodeError);
+      }
+
+      if (typeof fieldErrors !== "undefined") {
+        set(elementErrors, [nodeId, "fieldErrors"], fieldErrors);
+      }
```
Comment (BALEHOK) [inline src/pageEditor/slices/editorSlice.ts:703]: This action can be split into 2 separate actions.

File: src/pageEditor/tabs/editTab/EditTab.tsx [hunk 6]
```diff
@@ -167,23 +149,14 @@ const EditTab: React.FC<{
...
-                      name="label"
-                      label="Extension Name"
-                    />
-                    {showVersionField && <ApiVersionField />}
-                    <UpgradedToApiV3 />
-                    <SchemaFieldContext.Provider value={devtoolFieldOverrides}>
-                      <EditorNode isLocked={isLocked} />
-                    </SchemaFieldContext.Provider>
-                  </>
+                  <FoundationNodeConfigPanel
```
Comment (BALEHOK) [inline src/pageEditor/tabs/editTab/EditTab.tsx:152]: This change is not directly relevant to the error state, it just will make the Formik-Redux sync initialisation look the same way for Foundation and other nodes.

File: src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx [hunk 7]
```diff
@@ -56,6 +53,56 @@ const EditorNodeConfigPanel: React.FC<{
...
+      ) {
+        console.log("setting error in Formik with timeout", {
+          blockFieldName,
+          nodeError: nodeError.fieldErrors,
+        });
+        setError(nodeError.fieldErrors);
+        setTouched(setNestedObjectValues(nodeError.fieldErrors, true), false);
+      }
+    }, 1000);
+  }, [nodeId]);
```
Comment (BALEHOK) [inline src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx:97]: This is the synchronization of the Redux-Formik error state.

File: src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx [hunk 8]
```diff
@@ -68,11 +115,15 @@ const EditorNodeConfigPanel: React.FC<{
     [outputDescription]
   );
 
+  const newBlockError = useSelector(selectActiveNodeError);
+
   return (
     <>
-      {blockError && (
+      {(newBlockError?.message ?? blockError) && (
```
Comment (BALEHOK) [inline src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx:122]: `blockError` to be removed completely

File: src/components/fields/schemaFields/widgets/TextWidget.tsx [hunk 9]
```diff
@@ -68,7 +71,19 @@ const TextWidget: React.VFC<SchemaFieldProps & FormControlProps> = ({
   focusInput,
   ...formControlProps
 }) => {
-  const [{ value, ...restInputProps }, , { setValue }] = useField(name);
+  const [{ value, ...restInputProps }, , { setValue }] = useField({
+    name,
+    validate(value) {
```
Comment (BALEHOK) [inline src/components/fields/schemaFields/widgets/TextWidget.tsx:76]: String template validation runs on the widget level.

File: src/errors/errorHelpers.ts [hunk 10]
```diff
@@ -202,8 +201,12 @@ export function getErrorMessage(
     }
   }
 
-  if (error instanceof ValidationError) {
-    return error.errors.join(". ");
+  // @ts-expect-error -- We're checking if errors is there and if it's an array
+  if (Array.isArray(error.errors)) {
```
Comment (BALEHOK) [inline src/errors/errorHelpers.ts:205]: More general rule than `instanceof ValidationError`

File: src/pageEditor/tabs/editTab/FoundationNodeConfigPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import ApiVersionField from "@/pageEditor/fields/ApiVersionField";
+import UpgradedToApiV3 from "@/pageEditor/tabs/editTab/UpgradedToApiV3";
+import useFlags from "@/hooks/useFlags";
+import { isInnerExtensionPoint } from "@/registry/internal";
+import devtoolFieldOverrides from "@/pageEditor/fields/devtoolFieldOverrides";
+import SchemaFieldContext from "@/components/fields/schemaFields/SchemaFieldContext";
+import { UnconfiguredQuickBarAlert } from "@/pageEditor/extensionPoints/quickBar";
+import { BaseExtensionPointState } from "@/pageEditor/extensionPoints/elementConfig";
+
+const FoundationNodeConfigPanel: React.FC<{
```
Comment (BALEHOK) [inline src/pageEditor/tabs/editTab/FoundationNodeConfigPanel.tsx:29]: This code is pulled from the EditTab

File: src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx [hunk 12]
```diff
@@ -46,8 +39,10 @@ const EditorNodeConfigPanel: React.FC<{
    */
   blockFieldName: string;
   blockId: RegistryId;
-  blockError: string;
-}> = ({ blockFieldName, blockId, blockError }) => {
+  nodeId: UUID;
+}> = ({ blockFieldName, blockId, nodeId }) => {
+  useNodeValidation(blockFieldName, nodeId);
```
Comment (BALEHOK) [inline src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx:44]: This thing syncs between Redux and Formik the error state of the node inputs (current editor form).

File: src/pageEditor/tabs/editTab/editorNodeLayout/EditorNodeLayout.tsx [hunk 13]
```diff
@@ -145,6 +149,7 @@ const EditorNodeLayout: React.FC<EditorNodeLayoutProps> = ({
   const [allBlocks] = useAllBlocks();
   const activeNodeId = useSelector(selectActiveNodeId);
   const traces = useSelector(selectExtensionTrace);
+  const errors = useSelector(selectErrorMap);
```
Comment (BALEHOK) [inline src/pageEditor/tabs/editTab/editorNodeLayout/EditorNodeLayout.tsx:152]: This is like pipeline map but for errors

File: src/pageEditor/uiState/uiStateTypes.ts [hunk 14]
```diff
@@ -51,10 +52,35 @@ type PipelineMapBlock = {
   parentNodeId: UUID | null;
 };
 
+type NamespacedError = {
+  namespace: string;
```
Comment (BALEHOK) [inline src/pageEditor/uiState/uiStateTypes.ts:56]: `namespace` is needed for different validators to set and clear their errors independently

File: src/pageEditor/uiState/uiStateTypes.ts [hunk 15]
```diff
@@ -51,10 +52,35 @@ type PipelineMapBlock = {
...
+  fieldErrors: FormikErrorTree;
+};
+
 /**
  * The map of pipeline blocks. The key is the instanceId of the block.
  */
-export type PipelineMap = Record<UUID, PipelineMapBlock>;
+export type PipelineMap = Record<UUID, BlockInfo>;
+
+export type ErrorMap = Record<UUID, BlockErrorInfo>;
```
Comment (BALEHOK) [inline src/pageEditor/uiState/uiStateTypes.ts:83]: `BlockInfo` and `BlockErrorInfo`, how do you like the type names?

File: src/pageEditor/validation/applyTraceInputError.test.ts [hunk 16]
```diff
@@ -134,23 +134,3 @@ test("applies input validation error", () => {
     'Instance type "object" is invalid. Expected "string", "number", "boolean".'
   );
 });
-
-test("sets unknown input error on the block level", () => {
```
Comment (BALEHOK) [inline src/pageEditor/validation/applyTraceInputError.test.ts:138]: "unknown input error" falls into all other trace errors that are handled by `tracesValidator`, not just the test but this whole functionality is removed from `applyTraceInputError`

File: src/pageEditor/validation/renderersValidator.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+export const MULTIPLE_RENDERERS_ERROR_MESSAGE =
+  "A panel can only have one renderer. There are one or more renderers configured after this brick.";
+export const RENDERER_MUST_BE_LAST_BLOCK_ERROR_MESSAGE =
+  "A renderer must be the last brick.";
+
+class RenderersValidator implements Validator {
+  static namespace = "renderers";
+
+  matcher = isAnyOf(
```
Comment (BALEHOK) [inline src/pageEditor/validation/renderersValidator.ts:39]: The validator is triggered only when one of these actions is dispatched

File: src/pageEditor/validation/tracesValidator.ts [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+  selectErrorMap,
+} from "@/pageEditor/slices/editorSelectors";
+import { selectTraceErrors } from "@/pageEditor/slices/runtimeSelectors";
+import { actions as editorActions } from "@/pageEditor/slices/editorSlice";
+import { Validator, ValidatorEffect } from "./validationTypes";
+
+class TracesValidator implements Validator {
+  static namespace = "traces";
+
+  actionCreator = runtimeSlice.actions.setExtensionTrace;
```
Comment (BALEHOK) [inline src/pageEditor/validation/tracesValidator.ts:33]: This one runs only when traces change

File: src/pageEditor/validation/useNodeValidation.ts [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    },
+    500
+  );
+
+  // On load, get the error state from Redux and push to Formik
+  useEffect(() => {
+    // FIXME: don't use setTimeout here
+    // Formik seems to erase the error set on mount, so we wait a bit
+    // Figure out what's going on an fix it
+    setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/pageEditor/validation/useNodeValidation.ts:54]: > where the main <Formik/> context is being rendered
Correct. Not ideal. I'm looking into this part.

File: src/pageEditor/uiState/uiStateTypes.ts [hunk 20]
```diff
@@ -51,10 +52,35 @@ type PipelineMapBlock = {
...
+  namespace: string;
+  message: string;
+};
+
+export type BlockErrorInfo = {
+  /**
+   * Namespaced errors related to the Node
+   * @deprecated
+   */
+  nodeErrors: Record<string, string>;
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/pageEditor/uiState/uiStateTypes.ts:65]: This is just a leftover, dropped, thanks

File: src/pageEditor/validation/validationTypes.ts [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+export type ValidatorEffect = ListenerEffect<
+  AnyAction,
+  unknown,
+  ThunkDispatch<unknown, unknown, AnyAction>
+>;
+export type Validator = {
+  actionCreator?: TypedActionCreator<any>;
+  matcher?: MatchFunction<AnyAction>;
+  effect: ValidatorEffect;
+};
```
[the remark this answers is not in the document]
Comment (BALEHOK) [reply src/pageEditor/validation/validationTypes.ts:33]: Nice one, we can use RootState directly in ValidatorEffect, it doesn't have to be generic

[the remark this answers is not in the document]
Comment (BLoe) [reply src/pageEditor/validation/validationTypes.ts]: Yep, if we don't need that flexibility then we should definitely just type it to `RootState` directly 👍

File: src/pageEditor/validation/renderersValidator.ts [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+      .map(([nodeId]) => nodeId as UUID);
+
+    const allBlocks = await blockRegistry.allTyped();
+    const nodesWithErrors = new Set<UUID>();
+    traversePipeline({
+      pipeline: activeElement.extension.blockPipeline,
+      visitPipeline({ pipeline, pipelinePath, parentNode }) {
+        const isRootPipeline = parentNode === null;
+
+        // Only run validation for root pipeline and document Brick sub pipeline
```
[the remark this answers is not in the document]
Comment (BLoe) [reply src/pageEditor/validation/renderersValidator.ts:75]: Ah ok makes sense. Maybe change the comment to: "Only root pipelines and document builder Brick sub pipelines support renders, so we only need to run renderer validation on these types." Or even just the first part, before the comma?

File: src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx [hunk 23]
```diff
@@ -58,9 +58,12 @@ const EditorNodeConfigPanel: React.FC<{
...
-  const outputKeyLabel = useMemo(
-    () => <PopoverOutputLabel description={outputDescription} />,
-    [outputDescription]
+  const PopoverOutputLabel = (
+    <PopoverInfoLabel
+      name="output-label"
+      label="Output"
+      description={outputDescription}
+    />
   );
```
Comment (BLoe) [inline src/pageEditor/tabs/editTab/editorNodeConfigPanel/EditorNodeConfigPanel.tsx:67]: 👍

**Review by BLoe — APPROVED** [review]

**Pull request merged**